### PR TITLE
Implement BSD sendmsg()

### PR DESCRIPTION
**Pull request by jukkar**
Create implementation and unit tests for BSD socket sendmsg() API.

Also as an example, add SO_TXTIME socket option that uses the sendmsg() ancillary data to pass data from user space to network device driver.

### CODE REVIEW

File: include/net/net_pkt.h [hunk 1]
```diff
@@ -98,6 +98,10 @@ struct net_pkt {
 #if defined(CONFIG_NET_PKT_TIMESTAMP)
 	/** Timestamp if available. */
 	struct net_ptp_time timestamp;
+#endif
+#if defined(CONFIG_NET_PKT_TXTIME)
+	/** Network packet TX time in the future (in nanoseconds) */
+	u64_t txtime;
```
Comment (jukkar) [inline include/net/net_pkt.h]: This can probably combined with the timestamp field so we can save some space, I will rework this part.

**Comment by jukkar** [comment]
Fixed merge conflicts.

File: subsys/net/ip/net_context.c [hunk 2]
```diff
@@ -1131,10 +1131,36 @@ int net_context_get_timestamp(struct net_context *context,
 }
 #endif /* CONFIG_NET_CONTEXT_TIMESTAMP */
 
+/* If buf is not NULL, then use it. Otherwise read the data to be written
+ * to net_pkt from msghdr.
+ */
+static int write_data(struct net_pkt *pkt, const void *buf, int buf_len,
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: add context_ as a prefix to the function.

File: subsys/net/ip/net_context.c [hunk 3]
```diff
@@ -1260,15 +1287,34 @@ static int context_sendto(struct net_context *context,
...
+	if (!msghdr && !dst_addr &&
 	    !(IS_ENABLED(CONFIG_NET_SOCKETS_CAN) &&
 	      net_context_get_ip_proto(context) == CAN_RAW)) {
 		return -EDESTADDRREQ;
 	}
 
 	if (IS_ENABLED(CONFIG_NET_IPV6) &&
 	    net_context_get_family(context) == AF_INET6) {
-		struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *)dst_addr;
+		const struct sockaddr_in6 *addr6 =
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: could you make that change in a separate patch? same for addr4 below. (both in the same patch)

File: subsys/net/ip/net_context.c [hunk 4]
```diff
@@ -1276,7 +1309,7 @@ static int context_sendto(struct net_context *context,
 			  void *user_data,
 			  bool sendto)
 {
-	struct msghdr *msghdr = NULL;
+	const struct msghdr *msghdr = NULL;
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: move that change into the patch that introduced sendmsg

File: subsys/net/ip/net_context.c [hunk 5]
```diff
@@ -1582,6 +1629,7 @@ int net_context_send(struct net_context *context,
 
 int net_context_sendmsg(struct net_context *context,
 			const struct msghdr *msghdr,
+			size_t len,
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: same here

File: subsys/net/ip/net_context.c [hunk 6]
```diff
@@ -1512,6 +1580,24 @@ int net_context_send(struct net_context *context,
 	return ret;
 }
 
+int net_context_sendmsg(struct net_context *context,
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: len is missing (you added it in txtime patch, rebase was wrong I guess).

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (jukkar) [reply subsys/net/ip/net_context.c]: Hmm, not sure why we would need to separate this one. Now this is also bisectable.

File: subsys/net/ip/net_context.c (hunk 5 above)
Comment (jukkar) [reply subsys/net/ip/net_context.c]: Actually I removed this one, the len option is not needed.

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (tbursztyka) [reply subsys/net/ip/net_context.c]: because that has nothing really directly related to current patch but instead is a fix for an existing issue

Comment (jukkar) [reply subsys/net/ip/net_context.c]: Ah, you meant the const stuff, sure I can do that separately. I though you referred to other changes around these lines.

Comment (pfalcon) [reply subsys/net/ip/net_context.c]: > Ah, you meant the const stuff, sure I can do that separately.

+1, thanks @tbursztyka for suggesting that.

Comment (jukkar) [reply subsys/net/ip/net_context.c]: I separated the const thingy and after analysing the change, I am not sure this was really needed after all as we were not really fixing anything here.

**Review by tbursztyka — APPROVED** [review]
